### ISSUE TITLE
Awaits added to Users trigger twice.

### ISSUE THREAD

**Issue by W0lf619** [meew0/discordrb#430]
I was testing code the example for await to understand how it worked, specifically the number guessing game, and discovered that the await event would fire twice each time false was returned. So I would enter a number, and the bot would respond with it being either too high or too low twice, until I guessed correctly.

Here is a code snippet of what I was testing:
```ruby
# Start the game by typing "!game" in chat.
bot.message(start_with: '!game') do |event|
  # Pick a number between 1 and 10
  magic = rand(1..10)

  # Await a MessageEvent specifically from the invoking user.
  #
  # Note that since the identifier I'm using here is `:guess`,
  # only one person can be playing at one time. You can otherwise
  # interpolate something into a symbol to have multiple awaits
  # for this "command" available at the same time.
  event.user.await(:guess) do |guess_event|
    # Their message is a string - cast it to an integer
    guess = guess_event.message.content.to_i

    # If the block returns anything that *isn't* `false`, then the
    # event handler will persist and continue to handle messages.
    if guess == magic
      # This returns `nil`, which will destroy the await so we don't reply anymore
      guess_event.respond 'you win!'
    else
      # Let the user know if they guessed too high or low.
      guess_event.respond(guess > magic ? 'too high' : 'too low')

      # Return false so the await is not destroyed, and we continue to listen
      false
    end
  end

  # Let the user know we're  ready and listening..
  event.respond 'Guess a number between 1 and 10..'
end
```

**Comment by z64** [comment]
Hi there @W0lf619 !

Sorry - I can't reproduce this; copy and pasted the code from your issue.

![image](https://user-images.githubusercontent.com/6119032/32815456-2d4b50c0-c982-11e7-8739-f250993a4908.png)

Are you sure you weren't running two instances of the bot by accident?

**Comment by W0lf619** [comment]
I'm sure. Though I believe I've figured out the conditions that causes it to send twice. I test my commands by PMing my bot. Seems that if the bot is listening for an AwaitEvent, that event registers twice if it's in a PM.

![await bug](https://user-images.githubusercontent.com/32836691/32832605-f3a970c8-c9f3-11e7-9fcb-1bb79e0af274.png)

Outside of a PM, in a regular channel on a server, the event is only registered once, as your testing shows.

**Comment by z64** [comment]
Thank you for reporting. This makes sense as to why it's happening, but I'm very unsure how one would fix this. We'll take a closer look into it. :+1: May be slightly related to #422

**Comment by W0lf619** [comment]
Having done some more testing, it is indeed related to #422: Since a PrivateMessageEvent is a MessageEvent, the Await is firing off twice. Similar bug occurs if I mention someone while the Await is active.

**Comment by z64** [comment]
#422 can be fixed by doing:

```diff
diff --git a/lib/discordrb/events/reactions.rb b/lib/discordrb/events/reactions.rb
index d0e66b51..66dbd586 100644
--- a/lib/discordrb/events/reactions.rb
+++ b/lib/discordrb/events/reactions.rb
@@ -50,7 +50,7 @@ module Discordrb::Events
   class ReactionEventHandler < EventHandler
     def matches?(event)
       # Check for the proper event type
-      return false unless event.is_a? ReactionEvent
+      return false unless event.instance_of?(self.class)

       [
         matches_all(@attributes[:emoji], event.emoji) do |a, e|
```
However, I was anxious to apply this fix as I am unsure if it would break any other intended behavior.

**Comment by z64** [comment]
Well - that's what I wrote down however long ago I was looking into this. It doesn't work though, of course. This will be a bit more complicated :disappointed:

**Comment by LikeLakers2** [comment]
@z64 My initial thought on how to fix something like this is probably something like `return false unless event.class == self.class` in the same place as your suggested edit.

**Comment by z64** [comment]
meew suggested a different fix in the chat; I'll push it in a few, just testing it.

**Comment by z64** [comment]
@W0lf619 (Now that I've pushed the right branch) #431 should fix this; I tested both the reactions case and the example code here and they both work correctly.

From your issue I guess that you might not have any other implementations to test, but if you do, feel free to test / give any additional feedback in #431 .

**Comment by W0lf619** [comment]
@z64 Alrighty, will do once I get a chance to. 👍

**Comment by z64** [comment]
Should be fixed with #431 . Any additional issues like this, feel free to reopen or otherwise get in touch with us :+1:
